scripts/operations/system_layout_service/add_cdu_switch.py

```diff
@@ -36,32 +37,87 @@ def find_subnet(sls_network, name):
         existing_ip_reservations.add(ip_reservation["IPAddress"])
 
-    for available_ip in list(subnet[1:-2]):
+    start_ip = None
+    if len(ip_reservation) > 0:
+        start_ip = list(existing_ip_reservations)[0]
+
+    existing_ip_reservations.add(sls_subnet["Gateway"])
+
+    # Add reserved IP addresses. These included IP addresses that could be selected for DHCP.
+    for ip_reservation in reserved_ips:
+        existing_ip_reservations.add(ip_reservation)
+
+    index = 2  # The default is 2 to exclude the Gateway and the address after the gateway
+    if start_ip:
+        # find the index of the start_ip in the full list of IPs for the CIDR
+        i = 1
+        for ip in list(subnet[0:-2]):
+            if str(ip) == str(start_ip):
+                index = i
+                break
+            i += 1
+
+    # Find an IP that is not used starting at the index found above
+    for available_ip in list(subnet[index:-2]):
         if available_ip not in existing_ip_reservations:
             print("  {} Available for use.".format(available_ip))
             return available_ip
 
+    # If an IP wasn't found after the current reservations,
+    # then look for one in the IPs that come before the current reservations
+    if 1 < index:
+        for available_ip in list(subnet[1:index-1]):
+            if available_ip not in existing_ip_reservations:
+                print("  {} Available for use.".format(available_ip))
+                return available_ip
+
+def get_reserved_ips(sls_subnet):
+    name = sls_subnet["Name"]
+    reserved = []
+    if "DHCPEnd" in sls_subnet and "DHCPStart" in sls_subnet:
+        start = ip_address(sls_subnet["DHCPStart"])
+        end = ip_address(sls_subnet["DHCPEnd"])
+        if start <= end:
+            while start <= end:
+                print(f"  {name} Reserved for DHCP {start}")
+                reserved.append(str(start))
+                start += 1
+
+    for ip_reservation in sls_subnet.get("IPReservations", []):
+        ip = ip_reservation["IPAddress"]
+        print(f"  {name} Reserved {ip}")
+        reserved.append(ip)
+
+    return reserved
+
 def add_cdu_ip_reservation(sls_network, xname, alias):
+    print("Selecting IP Reservation for {} CDU Switch in {}'s network_hardware subnet".format(xname, sls_network["Name"]))
+
+    # collect all used IPs in every subnet of the subnet
+    reserved_ips = []
+    for subnet in sls_network["ExtraProperties"]["Subnets"]:
+        ips = get_reserved_ips(subnet)
+        reserved_ips.extend(ips)
+
     sls_subnet = find_subnet(sls_network, "network_hardware")
     if sls_subnet == None:
         print("Error: Unable to find network_hardware subnet in {} network!".format(network_name))
         exit(1)
 
-    print("Selecting IP Reservation for {} CDU Switch in {}'s network_hardware subnet".format(xname, sls_network["Name"]))
 
-    ip = find_next_available_ip(sls_subnet)
+    ip = find_next_available_ip(sls_subnet, reserved_ips=reserved_ips)
```

Comment:
There's no check to ensure a valid ip was returned from find_next_available_ip().  I presume that's because that is pretty much an impossible end case?

Reply:
I updated this